Quantized Conv2d Module

z-a-f opened the pull request:
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #21808 Quantized conv avoid functional usage&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15835572/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#21323 Quantized Conv2d Module**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15551835/)

Quantized Conv2d Module

Differential Revision: [D15551835](https://our.internmc.facebook.com/intern/diff/D15551835/)

Commit:
V8: Initial commit

Differential Revision: D15544061
Differential Version: 83630294

Commit:
V9: (no description)

Differential Revision: D15544061
Differential Version: 84141756

Commit:
V10: Added prepacked test

Differential Revision: D15544061
Differential Version: 84148224

Commit:
V2: Initial commit

Differential Revision: D15551835
Differential Version: 84162108

Commit:
V15: Rebase

Differential Revision: D15544061
Differential Version: 84413099

Commit:
V6: Rebase

Differential Revision: D15551835
Differential Version: 84413287

raghuramank100 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -0,0 +1,93 @@
+        self.groups = groups
+        self.padding_mode = padding_mode
+
+        weight = ops.quantized.fbgemm_conv_prepack(weight, groups)
```
Please use _packed as a prefix for packed tensors. 
self._packed_weight = weight

raghuramank100 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -0,0 +1,93 @@
+
+    @weak_script_method
+    def forward(self, input):
+        return qF.conv2d(input=input,
```
Instead of functional, having fbgemm calls which directly operate on packed weights would be better. packed weights are not exposed to the end user.

zafartahirov replied:
Will do in the next diff

raghuramank100 reviewed (approved):
Please change self.weight to self._packed_weight

Commit:
V18: Rebased and changed the packed layout

Differential Revision: D15544061
Differential Version: 84885783

Commit:
V7: Added groups test

Differential Revision: D15551835
Differential Version: 84894966

Commit:
V19: Added groups check

Differential Revision: D15544061
Differential Version: 84895794

Commit:
V20: (no description)

Differential Revision: D15544061
Differential Version: 84896229

Commit:
V9: cleanup

Differential Revision: D15551835
Differential Version: 84896360

Commit:
V10: Renamed the weight variable

Differential Revision: D15551835
Differential Version: 84950453

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -0,0 +1,93 @@
+# coding=utf-8
```
Please change API to https://github.com/pytorch/pytorch/pull/21921

raghuramank100 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -0,0 +1,93 @@
+
+@weak_module
+class _ConvNd(Module):
+    def __init__(self, weight, bias, scale, zero_point, dtype,
```
Second Jerry's comment. Please change API to match that of linear. The constructor should take the same arguments as float, with extra arguments corresponding to quantization.

raghuramank100 reviewed (changes requested):
Please address requests in comments, thanks

Commit:
V12: Rebase

Differential Revision: D15551835
Differential Version: 85764338

zafartahirov commented:
@raghuramank100 Some of the requested changes are in the stacked diff

Commit:
V13: Rebase

Differential Revision: D15551835
Differential Version: 85923499

jerryzh168 commented on test/test_quantized_conv.py:
```diff
@@ -6,6 +6,7 @@
 import numpy as np
 import torch
 import torch.nn.quantized.functional as qF
+from torch.nn.quantized.modules.conv import _conv_output_shape
```
is `torch.nn.quantized.modules.conv` considered as public? should we access them through `torch.nn.quantized` namespace?

zafartahirov replied:
It is public. This is a general PyTorch (and python) structure -- we have `torch.nn.modules.conv` and `torch.quantized.modules.conv` available to public, as well as having shortcuts in the `torch.nn` and `torch.nn.quantized` respectively.

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -0,0 +1,109 @@
+
+
+@weak_module
+class Conv2d(_ConvNd):
```
should we change to common base class for nn._ConvNd and nnq._ConvNd as we discussed the other day?

zafartahirov replied:
Good idea! Will do!

zafartahirov replied:
Actually, now that I am thinking about it, I don't think we need a common base. Deleting of the parent variables works, and in the future we will not use the packed weights, so it should be fine. What do you think?

jerryzh168 reviewed (changes requested):
please address comments

Commit:
V15: (no description)

Differential Revision: D15551835
Differential Version: 86033319

Commit:
V16: (no description)

Differential Revision: D15551835
Differential Version: 86083041

Commit:
V17: Rebase

Differential Revision: D15551835
Differential Version: 86218413

Commit:
Rebase and removal of weak scripts.

Tested:

```
$> python -m unittest test_nn_quantized.ModuleAPITest.test_conv_api test_quantized_conv.FunctionalAPITest.test_conv_api -v
test_conv_api (test_nn_quantized.ModuleAPITest)
Tests the correctness of the conv module. ... ok
test_conv_api (test_quantized_conv.FunctionalAPITest)
Tests the correctness of the conv functional. ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.996s

OK
```

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -0,0 +1,104 @@
+    @property
+    def weight(self):
+        return self._packed_weight
+        # return torch.ops.quantized.fbgemm_conv_unpack(self._packed_weight)
```
uncomment?

jerryzh168 commented on test/test_nn_quantized.py:
```diff
@@ -1,7 +1,13 @@
 import torch
-import torch.nn.quantized.functional as F
 import torch.nn.quantized as nnq
+import torch.nn.quantized.functional as F
```
remove?

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -0,0 +1,104 @@
+                    * (dilation - 1)) / stride) + 2 * output_padding + 1
+
+
+class Conv2d(_ConvNd):
```
can you add `from_float` and serialization that's similar to `nn.quantized.Linear`?
can be in another two prs.

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -0,0 +1,104 @@
+        self.register_buffer('_packed_weight',
+                             torch.ops.quantized.fbgemm_conv_prepack(qweight, self.groups))
+        self.register_buffer('bias', qbias)
+        self.register_buffer('_scale', torch.Tensor([1]))
```
types for scale and zero_point should be torch.double and torch.long actually, also can you make sure linear.py is the same as well? Thanks

jerryzh168 reviewed (approved):
looks good with some comments.

Commit:
Added explicit tensor type for scale and zero_point; Cleaned contamination

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -0,0 +1,110 @@
+        self.register_buffer('_packed_weight',
+            torch.ops.quantized.fbgemm_conv_prepack(qweight, self.groups))
+        self.register_buffer('bias', qbias)
+        self.register_buffer('_scale', torch.Tensor([1.0]).to(torch.double))
```
can you do
`torch.tensor([out_zero_point], dtype=torch.long)`

jerryzh168 replied:
I noticed that modules in nn/modules/linear.py are using torch.Tensor could you change them too

jerryzh168 replied:
asked Jianyu to change that for linear.py

zafartahirov replied:
Double is a default for tensor constructors. I think that's why it doesn't matter in the `nn/modules`

bddppq commented:
Lint broken:
https://travis-ci.org/pytorch/pytorch/jobs/557075852
```
./torch/nn/quantized/__init__.py:2:37: W292 no newline at end of file
./torch/nn/quantized/modules/conv.py:56:13: E128 continuation line under-indented for visual indent
./test/test_nn_quantized.py:119:32: E226 missing whitespace around arithmetic operator
```

zafartahirov commented:
Oh snap!!!!
